### DISCUSSION TITLE
Python oauth implementation troubleshooting

### DISCUSSION THREAD

**Discussion by skimike02** [timdorr/tesla-api#258] [Q&A] [answered]
Following  the steps from @uhthomas [here](https://github.com/timdorr/tesla-api/issues/215#issuecomment-770142656), I came up with the following python3. It works until it gets to the MFA url, which just returns a blank list 
```
b'{\n  "data": []\n}'
```
instead of the JSON expected. MFA is set up for the account in question.

Any help appreciated.

```python
import os
import base64
import hashlib
import random
import string
import requests
from bs4 import BeautifulSoup

verifier_bytes = os.urandom(86)
challenge = base64.urlsafe_b64encode(verifier_bytes).rstrip(b'=')
challenge_bytes = hashlib.sha256(challenge).digest()
challenge_sum = base64.urlsafe_b64encode(challenge_bytes).rstrip(b'=')
state=''.join(random.choices(string.ascii_uppercase + string.digits, k=10))

authorize_url="https://auth.tesla.com/oauth2/v3/authorize"

params={
    'client_id':'ownerapi',
    'code_challenge':challenge_sum,
    'code_challenge_method':'S256',
    'redirect_uri':'https://auth.tesla.com/void/callback',
    'response_type':'code',
    'scope':'openid email offline_access',
    'state':state}

data={
    'identity': [email],
    'credential': [pw]}

s=requests.Session()

r=s.get(authorize_url,params=params,data=data)
soup = BeautifulSoup(r.content, 'html.parser')

for item in soup.find_all('input'):
    if item.get('name') not in data.keys():
        data[item.get('name')]=[item.get('value')]

headers = {'Content-type': 'application/x-www-form-urlencoded'}        
r=s.post(authorize_url,headers=headers,params=params)

mfa_uri='https://auth.tesla.com/oauth2/v3/authorize/mfa/factors'

r=s.get(mfa_uri,data={'transaction_id':data['transaction_id'][0]})
```

The last GET has 
```
 'url': 'https://auth.tesla.com/oauth2/v3/authorize/mfa/factors',
 'body': 'transaction_id=[transactionID]',
```
where [transactionID] is an 8 character string with no brackets. entering random strings for [transactionID] gets the same result.

**Comment by uhthomas** [comment]
Sorry, maybe I should clarify the request for MFA devices is using a URL query, not an encoded form.

**Reply by skimike02** [reply]
Thank you. I tried that too, with the same result:
```
'url': 'https://auth.tesla.com/oauth2/v3/authorize/mfa/factors?transaction_id=[transactionID]',
'body': None,
```
It looks like my post request before MFA was hitting an error 400, so that is probably part of the problem. Perhaps I misinterpreted "Encode the map to the request body as a URL encoded form." I took this to mean the map we had created went in the body, and the other details noted went in the URL (shown below), though I also tried putting all items in the body.
```

{'method': 'POST',
 'url': 'https://auth.tesla.com/oauth2/v3/authorize?client_id=ownerapi&code_challenge[challenge]&code_challenge_method=S256&redirect_uri=https%3A%2F%2Fauth.tesla.com%2Fvoid%2Fcallback&response_type=code&scope=openid+email+offline_access&state=[state]',
 'headers': {'User-Agent': 'python-requests/2.25.1', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Content-type': 'application/x-www-form-urlencoded', 'Cookie': 'tesla-auth.sid=[cookie]', 'Content-Length': '181'},
 '_cookies': <RequestsCookieJar[Cookie(version=0, name='tesla-auth.sid', value=[value], port=None, port_specified=False, domain='auth.tesla.com', domain_specified=False, domain_initial_dot=False, path='/', path_specified=True, secure=True, expires=1612244422, discard=False, comment=None, comment_url=None, rest={'HttpOnly': None}, rfc2109=False)]>,
 'body': 'identity=[email]&credential=[credential]&_csrf=[csrf]&_phase=authenticate&_process=1&transaction_id=[transaction_ID]&cancel=',
 'hooks': {'response': []},
 '_body_position': None}
```

**Comment by heumn** [comment, marked as answer]
I have a working non-2FA and 2FA support here if you need some tips for how to do it in Python

https://github.com/enode-engineering/tesla-oauth2

**Reply by baysinger** [reply]
Thanks for your code. It does indeed work, but I'm trying to understand it. Why do you need to strip the padding (rstrip(b"=")) off of all the base64 representations?

Edit: To answer my own question: That's just a spec. RFC 7636 specifies the base64 encode function as this:
```
     static string base64urlencode(byte [] arg)
     {
       string s = Convert.ToBase64String(arg); // Regular base64 encoder
       s = s.Split('=')[0]; // Remove any trailing '='s
       s = s.Replace('+', '-'); // 62nd char of encoding
       s = s.Replace('/', '_'); // 63rd char of encoding
       return s;
     }
```
I'm still not sure why they do that. But they do!

**Comment by fkhera89** [comment]
The Mfa breaks or basically returns empty list if you have no Mfa devices enabled. If the goal is to just get the api working I would skip Mfa cause it makes this problem pretty complex and tracking refresh tokens etc. 

I used @heumn example code and it worked great !

**Reply by skimike02** [reply]
Apparently mine was returning empty list because I was locked out of my account from too many failed login attempts (the result of too much testing). I thought it would give a 401, but apparently it gives a 200 with a blank list.

**Comment by jeffkowalski** [comment]
I've added a ruby implementation of the full flow, based on @heumn's work
https://github.com/enode-engineering/tesla-oauth2/pull/4

**Comment by natrlhy** [comment]
How does this work if I have 2 MFA devices (Bitwarden and Duo) to select for the one time passcode?

**Reply by skimike02** [reply]
This code just takes the "first" device. I have made some modifications to the following that "should" (I only have one MFA device to test) ask in the console for you to select which device you want to enter the code for. To run, use authenticate(email,pw), then enter the number of your MFA device and your MFA code when prompted.

Please let me know if it works.

```python
import os
import base64
import hashlib
import requests
import time
from lxml import html
from urllib.parse import parse_qs


CLIENT_ID = "81527cff06843c8634fdc09e8ac0abefb46ac849f38fe1e431c2ef2106796384"
UA = "Mozilla/5.0 (Linux; Android 10; Pixel 3 Build/QQ2A.200305.002; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/85.0.4183.81 Mobile Safari/537.36"
X_TESLA_USER_AGENT = "TeslaApp/3.10.9-433/adff2e065/android/10"
authorize_url="https://auth.tesla.com/oauth2/v3/authorize"
token_url="https://auth.tesla.com/oauth2/v3/token"

def handshake():
    verifier_bytes = os.urandom(86)
    challenge = base64.urlsafe_b64encode(verifier_bytes).rstrip(b'=')
    challenge_bytes = hashlib.sha256(challenge).digest()
    challenge_sum = base64.urlsafe_b64encode(challenge_bytes).rstrip(b'=')
    state=base64.urlsafe_b64encode(os.urandom(32)).rstrip(b"=").decode("utf-8")
    return challenge, challenge_sum, state

def authenticate(email,pw):
    challenge, challenge_sum, state=handshake()
    
    MAX_ATTEMPTS = 10

    headers = {
        "User-Agent": UA,
        "x-tesla-user-agent": X_TESLA_USER_AGENT,
        "X-Requested-With": "com.teslamotors.tesla",
    }

    params={
        'client_id':'ownerapi',
        'code_challenge':challenge_sum,
        'code_challenge_method':'S256',
        'redirect_uri':'https://auth.tesla.com/void/callback',
        'response_type':'code',
        'scope':'openid email offline_access',
        'state':state}
    
    #Establish the session
    s=requests.Session()
    for i in range(MAX_ATTEMPTS):
        r=s.get(authorize_url,headers=headers,params=params)
        if r.ok and "<title>" in r.text:
            print(f"GET authorization form successful after {i + 1} attempt(s).")
            break
        time.sleep(3)
    else:
        raise ValueError(f"GET authorization form unsuccessful after {MAX_ATTEMPTS} attempts.")
                
    data = {
        "identity": email,
        "credential": pw,
    }
    
    tree=html.fromstring(r.content)
    for i in tree.xpath('//input'):
        if i.name not in data.keys() and i.type=='hidden':
            data[i.name]=i.value
            
    #Send email and pw
    for i in range(MAX_ATTEMPTS):
        r=s.post(authorize_url, headers=headers, data=data, params=params, allow_redirects=False)
        if r.ok and "<title>" in r.text:
            print(f"POST authorization form success after {i + 1} attempt(s).")
            break
        time.sleep(3)
    else:
        raise ValueError(f"POST authorization form unsuccessful after {MAX_ATTEMPTS} attempts.")
        
    #determine if MFA needed
    if r.status_code == 200 and "/mfa/verify" in r.text :
        mfa = True 
    else:
        mfa = False
    
    if mfa==True:
        mfa_uri='https://auth.tesla.com/oauth2/v3/authorize/mfa/factors'
        transaction_ID=data['transaction_id']
        r=s.get(mfa_uri, headers=headers, params={'transaction_id':transaction_ID})
        
        #Select MFA device if multiple exist:
        if len(r.json()['data'])>1:   
            print("Multiple MFA devices found. Please select which MFA device you will use.")
            n=1
            for i in r.json()['data']:
                print(n,"-",i['name'])
                n=n+1
                device=input()-1
        else:
            device=0
        factor_id=r.json()['data'][device]['id']
        
        #Send passcode
        for i in range(3):
            print('Please enter your passcode:')
            passcode=input()
            r = s.post("https://auth.tesla.com/oauth2/v3/authorize/mfa/verify",
                          headers=headers,
                          json={"transaction_id": transaction_ID, "factor_id": factor_id, "passcode": passcode})
            if r.json()['data']['approved'] and r.json()['data']['valid'] :
                print('passcode approved')
                break
        else:
            raise ValueError("Incorrect passcode entered")
        
        #Get code
        for attempt in range(MAX_ATTEMPTS):
            r = s.post(authorize_url, headers=headers, params=params, data=data, allow_redirects=False)
            if r.headers.get("location"):
                print(f"Got location in {attempt + 1} attempt(s).")
                break
            time.sleep(3)
        else:
            raise ValueError(f"Didn't get location in {MAX_ATTEMPTS} attempts.")
        
    else:
        print('No MFA detected')
    code = parse_qs(r.headers["location"])["https://auth.tesla.com/void/callback?code"]
    #Get access token
    payload = {
    "grant_type": "authorization_code",
    "client_id": "ownerapi",
    "code_verifier": challenge.decode("utf-8"),
    "code": code,
    "redirect_uri": "https://auth.tesla.com/void/callback",
        }
    
    r = s.post(token_url, headers=headers, json=payload)
    access_token = r.json()["access_token"]
    refresh_token=r.json()["refresh_token"]
            
    #Provide access token to get long-lived token
    headers["authorization"] = "bearer " + access_token
    payload = {
        "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
        "client_id": CLIENT_ID,
    }
    r = s.post("https://owner-api.teslamotors.com/oauth/token", headers=headers, json=payload)
    resp=r.json()
    resp['refresh_token']=refresh_token #refresh token needs to be from auth.tesla.com. owner-api refresh token is not used.
    return resp

def refresh(refresh_token):
    headers = {
        "User-Agent": UA,
        "x-tesla-user-agent": X_TESLA_USER_AGENT,
        "X-Requested-With": "com.teslamotors.tesla"}
    
    payload = {
    "grant_type": "refresh_token",
    "client_id": "ownerapi",
    "refresh_token": refresh_token,
    "scope": 'openid email offline_access'
        }
    r = requests.post(token_url, headers=headers, json=payload)
    access_token = r.json()["access_token"]
    refresh_token=r.json()["refresh_token"]
            
    #Provide access token to get long-lived token
    headers["authorization"] = "bearer " + access_token
    payload = {
        "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
        "client_id": CLIENT_ID,
    }
    r = requests.post("https://owner-api.teslamotors.com/oauth/token", headers=headers, json=payload)
    resp=r.json()
    resp['refresh_token']=refresh_token
    return resp
```

**Reply by natrlhy** [reply]
Do I hardcode my info here?

`
    data = {
        "identity": email,
        "credential": pw,
    }
`

**Reply by skimike02** [reply]
you could run the whole code, then run the command 
```
authenticate(email,pw)
```
replacing "email" with your email address in quotation marks, and "pw" with your password in quotation marks, like so:
```
authenticate('username@domain.com','sUpErSeCuRePaSsWoRd')
```
In response, you will get a JSON object that has the long-lived access token, creation timestamp, expires_in seconds, and refresh token.

**Reply by natrlhy** [reply]
the last line of code causes issues:
`return resp    ```
`
SyntaxError: invalid syntax

**Reply by skimike02** [reply]
> the last line of code causes issues:
> `` return resp ``` ``
> SyntaxError: invalid syntax

sorry. there was some markdown formatting left over. Just delete those last three punctuation marks (```). I edited the code in the comment above to get rid of them as well.

**Reply by natrlhy** [reply]
So I'm just getting into python, but when I edit the code in vs code I see this. I hate struggling with understanding it all so please, bear with me ;)

![image](https://user-images.githubusercontent.com/19696424/106938878-6182e400-66d4-11eb-8ea7-18b19e6a35b2.png)

**Reply by skimike02** [reply]
No worries.

In the "def" statement, you are listing the parameters (email and pw) that are the referenced later in that function. So you should leave that part as it was written in the code snippet. 

Run the whole file as written so that python knows what the function "authenticate(email,pw) means. Then, there should be a console where you enter commands one line at a time. There, you can run the command:
```
authenticate('youremail@here.com','yourpassword')
```
This tells python to run the authenticate function with 'youremail@here.com' replacing email and 'yourpassword' replacing pw.

Another way to do it would be at the very bottom of the file, unindented, to add the following code:
```python
creds=authenticate('youremail@here.com','yourpassword')
print(creds)
```
Then run the whole file and the console should print your credentials.

**Reply by natrlhy** [reply]
Thanks I think I'm used to just running the python script from cli with "python tesla.py" so should I be running this all right in the interpreter?

When I take your suggestion of putting in the creds code, it starts to work, but I get:
```
PS C:\Users\myuser> & C:/Users/myuser/AppData/Local/Microsoft/WindowsApps/python.exe c:/Users/myuser/Desktop/tesla_auth.py
GET authorization form successful after 2 attempt(s).
POST authorization form success after 2 attempt(s).
Multiple MFA devices found. Please select which MFA device you will use.
1 - Bitwarden
1
Traceback (most recent call last):
  File "c:\Users\myser\Desktop\tesla_auth.py", line 174, in <module>
    creds=authenticate('removed')
  File "c:\Users\myser\Desktop\tesla_auth.py", line 93, in authenticate
    device=input()-1
TypeError: unsupported operand type(s) for -: 'str' and 'int'
```

**Reply by skimike02** [reply]
yep. that's a bug. That's what you get when you don't test. Replace line 93 with the following and give it a whirl
```
device=int(input())-1
```

**Reply by natrlhy** [reply]
Here's my results. It did see multiple MFA devices. I input 1 and it asked again, showing the 2nd input method only after entering input method 1. I got my one time passcode and the exited out. I don't know where the auth code is :)

```
Multiple MFA devices found. Please select which MFA device you will use.
1 - Bitwarden
1
2 - Duo
1
Please enter your passcode:
838785
passcode approved
Got location in 3 attempt(s).
```

**Reply by skimike02** [reply]
oops. unindent line 93 one level so it shows you all the devices before you pick one, like so:
```python
        #Select MFA device if multiple exist:
        if len(r.json()['data'])>1:   
            print("Multiple MFA devices found. Please select which MFA device you will use.")
            n=1
            for i in r.json()['data']:
                print(n,"-",i['name'])
                n=n+1
            device=int(input())-1
        else:
            device=0
        factor_id=r.json()['data'][device]['id']
```
As for where the auth code is, when you run the command, it should return a json object. Not sure how that works in VS Code. You could just replace line 173 with ```print(resp)``` or ```print(resp['access_token'])```  instead of ```return resp```

**Reply by natrlhy** [reply]
Thanks. I've tried that but I'm still not getting any visual output of the token to use. Ideally I'd like to get this working on this:

https://github.com/natrlhy/teslaPowerwall

**Reply by skimike02** [reply]
oh, oops. swap line 146 also. 173 is for the refresh token process.

**Reply by natrlhy** [reply]
Yep that shows the auth_token! Thanks!

With the working code, I decided to remove my 2nd device auth method. Now it just sits:

```
GET authorization form successful after 2 attempt(s).
POST authorization form success after 1 attempt(s).
```

**Reply by skimike02** [reply]
That's odd, because I thought I tested it without MFA. maybe check your username/password? Currently the code doesn't look to see if they are correct.

You may want to try running the program piece by piece so you can see what the variable "r" is set to after you run line 67.
You could also add print statements to various areas to see where exactly it is hanging.

**Reply by Pyromancer579** [reply]
@Skimike02 Thank you for the code and the time and effort you've put in.

I'm trying to do much the same as @Natrhy, but want to swap between whatever mode my Powerwall is currently in to 'Back-Up' and visa versa as and when various conditions are met (to prevent my Powerwall discharging into my EV, when the EV charger conducts a timed charge).

I have slotted your code into mine (QuaterHourCheck.py - designed to be run by Cron every 15 minutes), so I can use my existing logic and conditions, and have put in a few print statements instead of the raise ValueError's, for the time being, so I can try and understand what is going on.

The best I can get is:

      GET authorization form successful after 1 attempt(s).
      POST authorization form unsuccessful after 10 attempts.
      No MFA Detected.

      Traceback (most recent call last):
      File "/home/pi/ZappiPowerwall/QuaterHourCheck.py", line 304, in <module>
      authenticate(TES_ID,TES_PW)
      File "/home/pi/ZappiPowerwall/QuaterHourCheck.py", line 161, in authenticate
      code = parse_qs(r.headers["location"])["https://auth.tesla.com/void/callback?code"]
      KeyError: 'https://auth.tesla.com/void/callback?code'

I've tried increasing MAX_ATTEMPTS to 20, to see if that would help, but no luck.

Any suggestions?

**Reply by dvermagithub** [reply]
First of thanks @skimike02 - you have saved me hours of sifting through XMLs and CSS. Not a front-end savvy person...so its greek to me.

@Pyromancer579  for line 68 
```
  #Send email and pw
    for i in range(MAX_ATTEMPTS):
        r=s.post(authorize_url, headers=headers, data=data, params=params, allow_redirects=False)
        if r.ok and "<title>" in r.text:
```

change to 
```
 #Send email and pw
    for i in range(MAX_ATTEMPTS):
        r=s.post(authorize_url, headers=headers, data=data, params=params, allow_redirects=False)
        if r.ok and r.status_code == 302:
```
or better yet

```
 #Send email and pw
    for i in range(MAX_ATTEMPTS):
        r=s.post(authorize_url, headers=headers, data=data, params=params, allow_redirects=False)
        if r.ok and (r.status_code == 302 or "<title>" in r.text):
```

**Reply by Pyromancer579** [reply]
@dvermagithub Thank you.

Got it working once I spotted the missing speach mark  after **<title>** in the 'better yet' solution.

Now have a 'refresh_token' is not defined error, even though I can print the value of 'refresh token' so off to poke around to try and solve this, I may be back on here asking for help later - hopefully not.

Thanks again.

**Reply by dvermagithub** [reply]
Np. Fixed missing quote mark

**Reply by Pyromancer579** [reply]
The token retrieval/authentication seems to be working okay now - had to make a few variables global in the appropriate places.

But now my original code for changing the mode of my Powerwall isn't working, I've put the code below and the response I get:

    256    async def Boost():
    257        client = TeslaApiClient(TES_ID, TES_PW)
    258        energy_sites = await client.list_energy_sites()
    259        assert(len(energy_sites)==1)
    260        N_MODE = await energy_sites[0].get_operating_mode()
    261        if N_MODE==C_MODE:
    262            await energy_sites[0].set_operating_mode('backup')
    263            N_MODE = await energy_sites[0].get_operating_mode()
    264        await client.close()

  **File "/home/pi/ZappiPowerwall/QuaterHourCheck.py", line 258, in Boost
    energy_sites = await client.list_energy_sites()
  File "/home/pi/.local/lib/python3.7/site-packages/tesla_api/__init__.py", line 106, in list_energy_sites
    return [_class(self, products['energy_site_id']) for products in await self.get('products')]
  File "/home/pi/.local/lib/python3.7/site-packages/tesla_api/__init__.py", line 79, in get
    await self.authenticate()
  File "/home/pi/.local/lib/python3.7/site-packages/tesla_api/__init__.py", line 67, in authenticate
    expiry_time = timedelta(seconds=self.token['expires_in'])
KeyError: 'expires_in'**

Looks to me like some sort of conflict between the old authentication code and the new system i.e. I'm still using the old system to try and get the required data and switch modes, but I don't know the appropriate code for doing the same under the new system.

Any guidance on how to alter the Powerwall mode would be much appreciated.

**Comment by labdel** [comment]
Hi All,
htanks for the python code, I pull the code and I ran it in my local PC pycharm and it works great I am able to get the taoken and refresh token. Problem I have is when I run same .py file in my AWS EC2 ubuntu machine it never returns in this line.
r=s.get(authorize_url,headers=headers,params=params)

I appeciate your feedback.

Thanks i nadvance - Leo.

**Reply by skimike02** [reply]
What happens when you run it? Do you get an error? Did you import requests as r?

**Reply by labdel** [reply]
It never returns, no error is returned, the only way is to ad timeout, ie. r=s.get(authorize_url,headers=headers,params=params, timeout=5).
I have the import Requests, same as in the code

Thanks

**Reply by skimike02** [reply]
can you open up a python environment on the EC2 machine and run through line by line? How can you tell that is the line not working?

**Reply by labdel** [reply]
That is the line that kicks off in the traceback when I interrupt hte execution after a while:

File "teslaauthentication.py", line 60, in login
    resp = session.get("https://auth.tesla.com/oauth2/v3/authorize", headers=headers, params=params)
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 543, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 670, in urlopen
    httplib_response = self._make_request(
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 426, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 421, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/local/lib/python3.8/http/client.py", line 1322, in getresponse
    response.begin()
  File "/usr/local/lib/python3.8/http/client.py", line 303, in begin
    version, status, reason = self._read_status()
  File "/usr/local/lib/python3.8/http/client.py", line 264, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/local/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "/usr/local/lib/python3.8/ssl.py", line 1241, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/local/lib/python3.8/ssl.py", line 1099, in read
    return self._sslobj.read(len, buffer)
KeyboardInterrupt

**Reply by skimike02** [reply]
No idea. I would recommend trying the code in sections on the EC2 instance. and see if you get a response at all, and if so, what you get. You could also try just using requests to make sure you can get a response from the tesla url without any of the rest of the script, in case it is something about the network config of the instance.

**Reply by alexmaple** [reply]
This sounds very similar to the problem I am having - script also works locally but not on EC2. I also spun up a Digital Ocean server to check and it doesn't work there either. Did you get anywhere with this?

**Reply by labdel** [reply]
I worked with a Network admin who helpt to re-route the traffic so it is not coming from AWS. It worked so far. It seems that they are banninc traffic from AWS for the authentication.

**Comment by labdel** [comment]
Hi,
In this part of the code:

   #Send email and pw
    for i in range(MAX_ATTEMPTS):
        r=s.post(authorize_url, headers=headers, data=data, params=params, allow_redirects=False)
        if r.ok and "<title>" in r.text:
            print(f"POST authorization form success after {i + 1} attempt(s).")
            break
        time.sleep(3)
    else:
        raise ValueError(f"POST authorization form unsuccessful after {MAX_ATTEMPTS} attempts.")

if r.ok and "<title>" in r.text: is True but later in

code = parse_qs(r.headers["location"])["https://auth.tesla.com/void/callback?code"]

I am not getting any 'location' in header.

Any suggestion?

Thank you.

**Reply by skimike02** [reply]
Does the account have MFA? If yes, there are many places it could have failed in between. 

If not, you might log into tesla website on a browser to make sure your password isn't locked out. I got locked out from sending too many bad requests, but the locked account still got http 200 codes. That confused me for a bit.

**Reply by Pyromancer579** [reply]
@skimike02 I'm having a problem in this area too - it manifested sometime between 7:30pm and 7:45pm on 23rd April 2021 (I don't know if that's relevant, but my last log entry was at 7:31pm that day).

It appears that my code now gets locked (in a loop?) at:

`r=s.post(authorize_url, headers=headers, data=data, params=params)`

I don't get any errors and the code doesn't stop, to narrow it down I put print statements before and after this line of code and running this indicates that the problem is in executing this line, as the prior print displays, but the code never gets to the next print statement on the following line.

I don't have MFA and I have logged into my Tesla account from a browser, to check that side of things and that seems to be fine, as is the Tesla app on my phone.

Any other ideas?

**Reply by skimike02** [reply]
yeah. they changed what user agent they accept. delete the line:    "User-Agent": UA, from the header definition and it should all start working again.

**Reply by Pyromancer579** [reply]
@skimike02  Thank you, that did the trick nicely.

**Reply by Pyromancer579** [reply]
@skimike02  Has Tesla changed something in their authentication systems again?

The python code I'm using worked fine up til and through 7th July 2021, but ceased working on the 8th!

**Reply by skimike02** [reply]
It looks like they added a captcha. Luckily I am using refresh tokens. I'm not sure if there is a way to bypass the captcha somehow...

**Reply by skimike02** [reply]
There is a discussion thread here: https://github.com/timdorr/tesla-api/discussions/390
I also modified my code here: https://github.com/skimike02/TeslaSolarData (you just need oauth.py) to download the captcha image and let you type it in.
